astropy/io/fits/tests/test_header.py

```diff
@@ -704,6 +704,22 @@ def test_hierarch_card_lookup(self):
         # should be treated case-insensitively when performing lookups
         assert 'ABCDEFGHI' in header
 
+    def test_hierarch_card_delete(self):
+        header = fits.Header()
+        header['hierarch abcdefghi'] = 10
+        del header['hierarch abcdefghi']
+
+    def test_hierarch_card_insert_delete(self):
+        header = fits.Header()
+        header['abcdefghi'] = 10
+        header['abcdefgh'] = 10
+        header['abcdefg'] = 10
+        header.insert(2, ('abcdefghij', 10))
+        del header['abcdefghij']
+        header.insert(2, ('abcdefghij', 10))
+        del header[2]
+        assert header.keys()[2]=='abcdefg'.upper()
```

Comment:
That's the reason why the tests failed. In python3 the keys() return a generator. You could cast it to a `list`:

```
assert list(header.keys())[2]=='abcdefg'.upper()
```